test/jdk/java/awt/Robot/MouseMoveOffScreen.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.awt.MouseInfo;
+import java.awt.Point;
+import java.awt.Robot;
+
+/*
+ * @test
+ * @bug 8345538
+ * @summary Tests mouseMove clamping to screen bounds when set to move offscreen
+ * @key headful
+ * @run main MouseMoveOffScreen
+ */
+
+public class MouseMoveOffScreen {
+    public static void main(String[] args) throws Exception {
+        Robot robot = new Robot();
+        robot.mouseMove(200, 200);
+        robot.delay(500);
+        robot.mouseMove(20000, 200);
+        robot.delay(500);
+
+        Point currLoc = MouseInfo.getPointerInfo().getLocation();
```

Comment:
Not directly related to this change, `MouseInfo.getPointerInfo().getLocation()` throws NPE for a multi-monitor extended display setup on windows for an off-screen location. It returns valid PointerInfo only if the point falls within the bounds of any of the screen device , else it returns null. This may be the expected behavior of MouseInfo.getPointerInfo() but the javadoc is not clear about how off-screen coordinates are handled.

For this test you can add a null check before calling .getLocation() for sanity (although multimonitor setup is not the case for CI runs)

Reply:
@aivanov-jdk 
I see your point. But I'm running into NPE without the fix as well. 

Is it expected of MouseInfo.getPointerInfo() to return null for off-screen coordinates (meaning does it consider it as mouse not available if it is outside the screen devices)?

A PointerInfo instance has two things associated with it - screen device and a location, so if it is not able to associate the coordinate with any of the screen device (off-screen coordinates) is it suppose to return null in this case?

Reply:
Interesting. Does this also happen after the fix with Robot clamping coordinates in shared code?

Reply:
Yes, I'm running into NPE without as well as with the fix.

Reply:
Is the current behaviour even a bug? [`Robot.mouseMove`](https://docs.oracle.com/en/java/javase/23/docs/api/java.desktop/java/awt/Robot.html#mouseMove(int,int)) says nothing about invalid coordinates.

> @aivanov-jdk I see your point. But I'm running into NPE at `MouseInfo.getPointerInfo().getLocation()` without the fix as well (dual monitor, extended display setup)

It's probably even better…

> Is it expected of MouseInfo.getPointerInfo() to return null for off-screen coordinates (meaning does it consider it as mouse not available if it is outside the screen devices)?
> 
> PointerInfo has two things associated with it - screen device and a location, so if it is not able to associate the coordinate with any of the screen device (off-screen coordinate) is it suppose to return null in this case?

So, the mouse seems to be positioned outside of the virtual screen bounds. For this reason, `MouseInfo.getPointerInfo()` cannot return anything.

This again raises the question, why does it return a non-null value on macOS?

What do you see on the screen? Does the mouse cursor or its part remain visible on the screen? What happens when you start moving your mouse?

Reply:
> Yes, I'm running into NPE without as well as with the fix.

Can you provide information on your multi-screen setup ?
how many displays ? what is the relative placement of the monitors ?
What are the virtual bounds of your setup ?

And the stack trace of the exception you get too.

Reply:
For me, the test *without fix* prints
```text
Current mouse location: java.awt.Point[x=0,y=200]
```
The mouse cursor moves to 200, 200 and then to 0, 200 on the screen.

Reply:
Displays: 2 
Placement: Side-by-side, with the bottom edge aligned as below

<img width="478" alt="image" src="https://github.com/user-attachments/assets/7ea5be69-989f-42b5-ac26-631d170d8db0" />

Screen bounds:
Win32GraphicsDevice[screen=0]:java.awt.Rectangle[x=-1920,y=363,width=1280,height=720]
Win32GraphicsDevice[screen=1]:java.awt.Rectangle[x=0,y=0,width=2293,height=960]

Stack Trace:
----------System.err:(11/959)----------
java.lang.NullPointerException: Cannot invoke "java.awt.PointerInfo.getLocation()" because the return value of "java.awt.MouseInfo.getPointerInfo()" is null
	at MouseMoveOffScreen.main(MouseMoveOffScreen.java:57)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1447)

JavaTest Message: Test threw exception: java.lang.NullPointerException: Cannot invoke "java.awt.PointerInfo.getLocation()" because the return value of "java.awt.MouseInfo.getPointerInfo()" is null
JavaTest Message: shutting down test

Reply:
> For me, the test _without the fix_ prints
> 
> ```
> Current mouse location: java.awt.Point[x=0,y=200]
> ```
> 
> The mouse cursor moves to 200, 200 and then to 0, 200 on the screen.


What platform ? Alisen did say in the description that Windows & Linux clamp which presumably is either something elsewhere in the JDK implementation, or just how the platform works. I looked (briefly) at Linux and we just pass the values to XWarpPointer so I assumed it was the platform ..

Reply:
In MouseInfo.getPointerInfo()
The point is updated here [fillPointWithCoords()] (https://github.com/openjdk/jdk/blob/7ba969a576eb92446e40587fecf98e1e4aba8883/src/java.desktop/share/classes/java/awt/MouseInfo.java#L72) .

and what follows after

```
for (int i = 0; i < gds.length; i++) {
                GraphicsConfiguration gc = gds[i].getDefaultConfiguration();
                Rectangle bounds = gc.getBounds();
                if (bounds.contains(point)) { // checks if the point updated by  fillPointWithCoords is within the bounds of any of the screen device else it returns null
                    retval = new PointerInfo(gds[i], point);
                }
            }
```

Reply:
I ran it on Windows, sorry I should've said it explicitly.

Reply:
> Displays: 2 Placement: Side-by-side, with the bottom edge aligned as below
> 
> <img alt="image" width="478" src="https://private-user-images.githubusercontent.com/95945681/397856615-7ea5be69-989f-42b5-ac26-631d170d8db0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ3Mjc0NjMsIm5iZiI6MTczNDcyNzE2MywicGF0aCI6Ii85NTk0NTY4MS8zOTc4NTY2MTUtN2VhNWJlNjktOTg5Zi00MmI1LWFjMjYtNjMxZDE3MGQ4ZGIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjIwVDIwMzkyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhYWUzZWNkZWZjOGFiZTRjMWE4ODgxMjhhYjE3N2E4NWRkOTA2MmI0MzMyZWJmNmExYTJmODA0ZjgzZGJmMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vGpCyq1K_5DGxJ-rZPPp5PosfvP4Q90l-3hIAWb1E7M">
> Screen bounds: Win32GraphicsDevice[screen=0]:java.awt.Rectangle[x=-1920,y=363,width=1280,height=720] Win32GraphicsDevice[screen=1]:java.awt.Rectangle[x=0,y=0,width=2293,height=960]
> 
> Stack Trace: ----------System.err:(11/959)---------- java.lang.NullPointerException: Cannot invoke "java.awt.PointerInfo.getLocation()" because the return value of "java.awt.MouseInfo.getPointerInfo()" is null at MouseMoveOffScreen.main(MouseMoveOffScreen.java:57) at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) at java.base/java.lang.reflect.Method.invoke(Method.java:565) at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138) at java.base/java.lang.Thread.run(Thread.java:1447)
> 
> JavaTest Message: Test threw exception: java.lang.NullPointerException: Cannot invoke "java.awt.PointerInfo.getLocation()" because the return value of "java.awt.MouseInfo.getPointerInfo()" is null JavaTest Message: shutting down test

The stack trace wasn't as revealing as I hoped. But the monitor config may tell us something.
(0, 200) is a location that is off the top of screen 1, so not within the virtual bounds.
But it is interesting that you get NPE before the fix as well .. perhaps the windows clamping isn't happening as Alisen said. I think we need to understand that as the next step.

After that , since it looks to me like the bounds of all screens need to be looked at to clamp because we can't rely on a single rectangle defining it even if there's a single virtual screen, and need to look at the union.

Reply:
Now I ran the test on macOS Sequoia 15.1.1. With the current `Robot` implementation, it's able to move mouse across both displays. With Alisen's updated implementation, the coordinates are limited to the main screen.

<details>
<summary>Updated test</summary>

```java
    private static void moveMouseAndCheck(final Robot robot,
                                          final int x, final int y) {
        System.out.println(x + ", " + y);
        robot.mouseMove(x, y);
        robot.delay(1000);

        PointerInfo cursor = MouseInfo.getPointerInfo();
        if (cursor != null) {
            System.out.println(cursor.getLocation() + " - "
                               + cursor.getDevice().getIDstring());
        }
    }

    public static void main(String[] args) throws Exception {
        Robot robot = new Robot();

        GraphicsDevice[] screens = getLocalGraphicsEnvironment()
                                   .getScreenDevices();
        for (GraphicsDevice screen : screens) {
            GraphicsConfiguration gc = screen.getDefaultConfiguration();
            System.out.println(screen.getIDstring());
            System.out.println("\t" + gc.getBounds());
            System.out.println("\t" + gc.getDefaultTransform());
        }
        System.out.println("\n");

        for (GraphicsDevice screen : screens) {
            System.out.println(screen.getIDstring());
            Rectangle bounds = screen.getDefaultConfiguration()
                                     .getBounds();

            moveMouseAndCheck(robot, bounds.x, bounds.y);
            moveMouseAndCheck(robot, bounds.x + bounds.width, bounds.y);
            moveMouseAndCheck(robot, bounds.x, bounds.y + bounds.height);
            moveMouseAndCheck(robot, bounds.x + bounds.width, bounds.y + bounds.height);
            moveMouseAndCheck(robot, bounds.x + bounds.width / 2, bounds.y + bounds.height / 2);
        }
```
</details>

Displays are arranged side by side: the built-in Display is on the left, the main display is on the right.

<details>
<summary>JDK 21 without the fix</summary>

**JDK 21 without the fix**
```text
Display 69734208
	java.awt.Rectangle[x=-1440,y=0,width=1440,height=900]
	AffineTransform[[2.0, 0.0, 0.0], [0.0, 2.0, 0.0]]
Display 725353101
	java.awt.Rectangle[x=0,y=0,width=1920,height=1080]
	AffineTransform[[2.0, 0.0, 0.0], [0.0, 2.0, 0.0]]


Display 69734208
-1440, 0
java.awt.Point[x=-1440,y=0] - Display 69734208
0, 0
java.awt.Point[x=0,y=0] - Display 725353101
-1440, 900
cursor == null
0, 900
java.awt.Point[x=0,y=900] - Display 725353101
-720, 450
java.awt.Point[x=-720,y=450] - Display 69734208
Display 725353101
0, 0
java.awt.Point[x=0,y=0] - Display 725353101
1920, 0
cursor == null
0, 1080
cursor == null
1920, 1080
cursor == null
960, 540
java.awt.Point[x=960,y=540] - Display 725353101
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "java.awt.PointerInfo.getLocation()" because the return value of "java.awt.MouseInfo.getPointerInfo()" is null
	at MouseMoveOffScreen.main(MouseMoveOffScreen.java:90)
```
</details>


<details>
<summary>With the fix</summary>

**With the fix**
```text
Display 69734208
	java.awt.Rectangle[x=-1440,y=0,width=1440,height=900]
	AffineTransform[[2.0, 0.0, 0.0], [0.0, 2.0, 0.0]]
Display 725353101
	java.awt.Rectangle[x=0,y=0,width=1920,height=1080]
	AffineTransform[[2.0, 0.0, 0.0], [0.0, 2.0, 0.0]]


Display 69734208
-1440, 0
2024-12-20 20:39:09.383 java[94658:1065578] +[IMKClient subclass]: chose IMKClient_Modern
2024-12-20 20:39:09.383 java[94658:1065578] +[IMKInputSession subclass]: chose IMKInputSession_Modern
java.awt.Point[x=0,y=0] - Display 725353101
0, 0
java.awt.Point[x=0,y=0] - Display 725353101
-1440, 900
java.awt.Point[x=0,y=900] - Display 725353101
0, 900
java.awt.Point[x=0,y=900] - Display 725353101
-720, 450
java.awt.Point[x=0,y=450] - Display 725353101
Display 725353101
0, 0
java.awt.Point[x=0,y=0] - Display 725353101
1920, 0
cursor == null
0, 1080
cursor == null
1920, 1080
cursor == null
960, 540
java.awt.Point[x=960,y=540] - Display 725353101
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "java.awt.PointerInfo.getLocation()" because the return value of "java.awt.MouseInfo.getPointerInfo()" is null
	at MouseMoveOffScreen.main(MouseMoveOffScreen.java:90)
```
</details>

In both cases, the points at the right and bottom edges of the screen aren't available, `MouseInfo.getPointerInfo()` returns `null`. I should've subtracted 1 when adding `bounds.width` or `bounds.height`.

Reply:
Note that even with Alisen's fix, **the test fails to get the mouse pointer location after moving it off the screen!**

The coordinates are expected to be within the screen bounds.

Reply:
> (0, 200) is a location that is off the top of screen 1, so not within the virtual bounds.

@prrace  Why is it? Harshitha's screen 1 bounds are: \[x=0,y=0,width=2293,height=960\], so (0, 200) lies within the bounds of the screen.

Alisen's test moves mouse to (200, 200) and then to (20_000, 200). I expected to see mouse move to the right edge of the (virtual) screen but mouse pointer is instead at (0, 200) for me.

I don't know where mouse pointer is in Harshitha's case.

On Windows, I have a second monitor to the right of my main one, and main one has scale of 2.0. Harshitha has her secondary monitor on the left, I presume. Where is the mouse cursor on the screen?

By looking at the numbers, Harshitha has scale on her screens. Scaling down what's received from Windows could trim a pixel on the right, and it could be the reason why the point isn't within the bounds of any screen in her case.

Reply:
@alisenchung @honkar-jdk @prrace Things align: if the main monitor is on the right, and the secondary monitor is to the left of the main one, `Robot.mouseMove(20_000, 200)`, the following call to `MouseInfo.getPointerInfo()` returns `null`.

Both Harshitha and I see `null` in such a configuration both with and without the fix… on Windows and Mac.

Reply:
@aivanov-jdk 
> Why is it? Harshitha's screen 1 bounds are: [x=0,y=0,width=2293,height=960], so (0, 200) lies within the bounds of the screen.

You are right, I'm able to see the mouse pointer on my [screen=1] and it lies within screen bounds with this configuration.

Left screen [screen=0]:java.awt.Rectangle[x=-1920,y=363,width=1280,height=720]
**[Primary]** Right side screen [screen=1]:java.awt.Rectangle[x=0,y=0,width=2293,height=960]

> I don't know where mouse pointer is in Harshitha's case.
> Harshitha has her secondary monitor on the left, I presume. Where is the mouse cursor on the screen?

Yes, I have the secondary screen on left and primary on right. And with Robot.mouseMove(200, 200) , the mouse cursor is on the primary screen (right hand-side)

Reply:
>  Things align: if the main monitor is on the right, and the secondary monitor is to the left of the main one, Robot.mouseMove(20_000, 200), the following call to MouseInfo.getPointerInfo() returns null.

@aivanov-jdk You are right. I tested by switching the primary monitor and the test passed. 

With Robot.mouseMove(20000, 200)

**Left side: Secondary  & Right side: Primary screen --> Test fails**
Win32GraphicsDevice[screen=0]:java.awt.Rectangle[x=-1920,y=363,width=1280,height=720]
**[Primary]** Win32GraphicsDevice[screen=1]:java.awt.Rectangle[x=0,y=0,width=2293,height=960]

**Left side: Primary  & Right: Secondary --> Test passes**
**[Primary]** Win32GraphicsDevice[screen=0]:java.awt.Rectangle[x=0,y=0,width=1280,height=720]
Win32GraphicsDevice[screen=1]:java.awt.Rectangle[x=1920,y=-363,width=2293,height=960]